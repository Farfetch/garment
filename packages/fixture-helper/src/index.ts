export * from './executeRunner';
export * from './initFixtureHelper';
export * from './resolver';
import normalizePath = require('normalize-path');
import isValidPath = require('is-valid-path');

const containsValidPath = (value: string): boolean =>
  value.split(/\s/).some((s) => isValidPath(s));

export const normalizeContainedPath = (value: string): string => {
  return value
    .split(/(\s)/)
    .map((s) => {
      if (isValidPath(s)) {
        return normalizePath(s);
      } else {
        return s;
      }
    })
    .join('');
};

export function replaceTestPath<T>(
  obj: T,
  testPath: string,
  replaceWith = '/test_path'
): T {
  const normalizedTestPathRegEx = new RegExp(normalizePath(testPath), 'g');
  return JSON.parse(
    JSON.stringify(obj, (_, value) =>
      typeof value === 'string' && containsValidPath(value)
        ? normalizeContainedPath(value).replace(
            normalizedTestPathRegEx,
            replaceWith
          )
        : value
    )
  );
}
