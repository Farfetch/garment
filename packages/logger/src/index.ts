import chalk from 'chalk';
import stripAnsi from 'strip-ansi';

const levelsMap = {
  silly: 0,
  debug: 1,
  info: 2,
  success: 3,
  warn: 4,
  error: 5,
  silent: 6,
};

export type Level = keyof typeof levelsMap;

const logTypeToTextMap: { [key in Level]?: string } = {
  silly: chalk.keyword('purple')('silly'),
  debug: chalk.keyword('cyan')('debug'),
  info: chalk.keyword('lightblue')('info'),
  warn: chalk.keyword('yellow')('warning'),
  error: chalk.keyword('darkred')('error'),
  success: chalk.keyword('lime')('success'),
};

export function shouldPrint(level: Level, currentLevel: Level) {
  return levelsMap[level] >= levelsMap[currentLevel];
}

export const levels = Object.keys(levelsMap);

export class Logger {
  public static print(
    options: {
      prefix?: string;
      scope?: string;
      level?: Level;
      scopeColor?: string;
      dontLog?: boolean;
    } = {},
    ...args: any[]
  ) {
    const meta: string[] = [];
    const { prefix, scope, level, scopeColor = 'magenta', dontLog } = options;
    if (prefix) {
      meta.push(chalk.grey(prefix));
    }
    if (scope) {
      meta.push(chalk.keyword(scopeColor)(scope));
    }
    if (level) {
      meta.push(logTypeToTextMap[level] || level);
    }
    const joinedMeta = meta.join(' ');

    const indentedArgs = args
      .map((arg) => (Object(arg) !== arg ? arg : JSON.stringify(arg)))
      .join(' ')
      .split(/\n/)
      .join('\n' + ' '.repeat(stripAnsi(joinedMeta).length + 1));

    const consoleLogArgs: string[] = [joinedMeta, indentedArgs];
    if (!dontLog) {
      console.log(...consoleLogArgs);
    }
    return consoleLogArgs;
  }

  interceptors: ((entry: { level: Level; args: any[] }) => any)[] = [];

  constructor(
    public readonly prefix: string | undefined,
    public readonly name: string,
    public level: Level = 'info',
    public readonly color?: string
  ) {}

  public silly(...args: any[]) {
    this.display('silly', ...args);
  }

  public debug(...args: any[]) {
    this.display('debug', ...args);
  }

  public info(...args: any[]) {
    this.display('info', ...args);
  }

  public warn(...args: any[]) {
    this.display('warn', ...args);
  }

  public error(...args: any[]) {
    this.display('error', chalk.keyword('red')(...args));
  }

  public success(...args: any[]) {
    this.display('success', ...args);
  }

  public log(...args: any[]) {
    Logger.print({}, ...args);
  }

  public createChild(scope: string, color?: string) {
    return new Logger(this.prefix, scope, this.level, color);
  }

  public setLevel(level: Level) {
    this.level = level;
  }

  private display(level: Level, ...args: any[]) {
    if (
      !this.interceptors.every((fn) => {
        const result = fn({ level, args });
        return typeof result === 'boolean' ? result : true;
      })
    ) {
      return;
    }
    if (shouldPrint(level, this.level)) {
      Logger.print(
        {
          prefix: this.prefix,
          scope: this.name,
          level,
          scopeColor: this.color,
        },
        ...args
      );
    }
  }
}

export default new Logger('', '', 'info');
