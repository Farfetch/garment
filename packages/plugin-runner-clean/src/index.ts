import { defineRunner } from '@garment/runner';
import del from 'del';

/**
 * Deletes files and folders using glob patterns.
 */
export interface CleanRunnerOptions {}

export default defineRunner(async function (ctx) {
  ctx.logger.info(`Removing files`);

  ctx.input.forEach(async ({ path, baseDir }) => {
    if (baseDir) {
      await del(path, {
        cwd: baseDir,
      });
    } else {
      throw new Error(`Unable to delete "${path}". baseDir is empty`);
    }
  });

  ctx.logger.success(`Files removed`);
});
