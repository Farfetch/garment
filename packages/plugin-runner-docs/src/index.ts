import { defineRunner, defineOptions } from '@garment/runner';
import * as Path from 'path';
import { Project, SourceFile } from 'ts-morph';

export default defineRunner(
  defineOptions({
    template: ['path', 'Path to the template'],
    filename: ['string?', 'Name of the output file', 'README.md']
  }),
  async ctx => {
    const { template, filename } = ctx.options;
    const packageName = ctx.project.nodePackage?.name;

    const templateContent = ctx.renderTemplate(
      ctx.fs.readFileSync(template, 'utf8'),
      { packageName }
    );

    const templateLines = templateContent.split('\n');
    const updatedTemplateLines = [];

    const project = new Project({});
    const sourceFiles: Record<string, SourceFile> = {};

    let ignoreLine = false;

    for (const line of templateLines) {
      const match = /<!--\s*(.*)\s*-->/.exec(line);
      if (match) {
        const [op, ...args] = match[1].split(/\s+/);

        if (op === 'api') {
          const [filename, declarationName] = args;

          const sourceFile = getSourceByFilename(filename);

          const declaration =
            declarationName === 'firstInterface'
              ? sourceFile.getInterfaces()[0]
              : sourceFile.getInterfaceOrThrow(declarationName);

          const allResults = [];

          for (const member of declaration.getMembers()) {
            let result = [];
            const name = (member as any).getName();
            let type = member.getType().getText();
            let defaultValue;
            let comment;

            const [jsdoc] = member.getJsDocs();
            if (jsdoc) {
              comment = jsdoc.getDescription().trim();
              const tags: Record<string, string> = {};
              for (const tag of jsdoc.getTags()) {
                tags[tag.getTagName()] = tag.getComment()?.trim() ?? '';
              }

              if (tags.format === 'path') {
                type = 'path';
              }

              defaultValue = tags.defaultDescription || tags.default;
            }

            result.push(`\`${name}: ${type}\``);
            if (defaultValue) {
              result.push(`**Default:** \`${defaultValue}\``);
            }
            if (comment) {
              result.push(comment);
            }

            allResults.push(result.join('\n\n'));
          }

          ignoreLine = true;

          updatedTemplateLines.push(line);

          allResults.forEach((result, index) => {
            updatedTemplateLines.push(result);
            if (index === allResults.length - 1) {
              updatedTemplateLines.push('\n');
            } else {
              updatedTemplateLines.push('\n---\n');
            }
          });
        } else if (op === 'example' || op === 'description') {
          const [filename, declarationName] = args;

          const sourceFile = getSourceByFilename(filename);

          const declaration =
            declarationName === 'firstInterface'
              ? sourceFile.getInterfaces()[0]
              : sourceFile.getInterfaceOrThrow(declarationName);

          const allResults = [];

          const [jsdoc] = declaration.getJsDocs();
          if (jsdoc) {
            if (op === 'example') {
              const tags: Record<string, string> = {};
              for (const tag of jsdoc.getTags()) {
                tags[tag.getTagName()] = tag.getComment()?.trim() ?? '';
              }

              if (tags.example) {
                let parsed = tags.example;
                try {
                  parsed = JSON.stringify(JSON.parse(parsed), null, 4);
                } catch (error) {}

                allResults.push('```json');
                allResults.push(parsed);
                allResults.push('```\n');
              }
            } else if (op === 'description') {
              const description = jsdoc.getDescription().trim();
              allResults.push(description);
              allResults.push('\n');
            }
          }

          ignoreLine = true;

          updatedTemplateLines.push(line, ...allResults);
        } else if (op === 'installation') {
          updatedTemplateLines.push(line);
          if (packageName) {
            updatedTemplateLines.push(`\`npm i ${packageName}\`\n`);
            updatedTemplateLines.push(`or\n`);
            updatedTemplateLines.push(`\`yarn add ${packageName}\`\n`);
          }

          ignoreLine = true;
        } else if (op === 'break') {
          ignoreLine = false;
        }
      } else if (line[0] === '#') {
        ignoreLine = false;
      }

      if (!ignoreLine) {
        updatedTemplateLines.push(line);
      }
    }

    return ctx.file.text(filename, updatedTemplateLines.join('\n'));

    function getSourceByFilename(filename: string) {
      const resolvedFilename = Path.resolve(ctx.project.fullPath, filename);
      if (!sourceFiles[resolvedFilename]) {
        sourceFiles[resolvedFilename] = project.addSourceFileAtPath(
          resolvedFilename
        );
      }
      return sourceFiles[resolvedFilename];
    }
  }
);
