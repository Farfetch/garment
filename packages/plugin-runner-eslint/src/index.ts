import {
  defineOptionsFromJSONSchema,
  defineRunner,
  File
} from '@garment/runner';
import { CLIEngine } from 'eslint';

/**
 * Runs ESLint
 * @example {
    "runner": "eslint",
    "input": "{{projectFolder}}/src",
    "options": {
      "configFile": ".eslintconfig.json"
    }
  }
 */
interface EslintRunnerOptions {
  /**
   * The config file to load ESLint's configuration from.
   * @format path
   */
  configFile?: string;

  /**
   * Fix automatically linting errors.
   */
  fix?: boolean;
}

enum SeverityLevel {
  Warning = 1,
  Error = 2
}

export default defineRunner(
  defineOptionsFromJSONSchema<EslintRunnerOptions>(require('./schema.json')),
  async function(ctx) {
    const { logger, options, input, workspace, project } = ctx;
    const { configFile, outputDir, fix: fixFlag } = options;
    const fix = Boolean(outputDir) && fixFlag;

    /*
     * Updates the cwd to be the project
     * In this way the TS parser will only parse the project files
     */
    process.chdir(project.fullPath);

    logger.info(`Linting files with ESLint`);

    const cliEngineOpts: CLIEngine.Options = { fix, cwd: workspace.cwd };
    // if user specifies configFile should set useEslintrc to false
    if (configFile) {
      cliEngineOpts.configFile = configFile;
      cliEngineOpts.useEslintrc = false;
    }

    const cli = new CLIEngine(cliEngineOpts);

    ctx.input(files => {
      const fixes: File[] = [];

      files.forEach(file => {
        const { results } = cli.executeOnText(
          file.data.toString('utf8'),
          file.absolutePath
        );

        const warningResults = filterResultsBySeverity(
          results,
          SeverityLevel.Warning
        );
        const errorResults = filterResultsBySeverity(
          results,
          SeverityLevel.Error
        );
        const formatter = cli.getFormatter();
        const regexMatcher = /\n.*\u2716 .*\).*\n/;
        const fixableMessageMatcher = /[0-9]+ errors and [0-9]+ warnings .*/;
        if (errorResults.length) {
          const errorPrint = formatter(errorResults)
            .replace(regexMatcher, '')
            .replace(fixableMessageMatcher, '');
          logger.error(errorPrint);
        }
        if (warningResults.length) {
          logger.warn(
            formatter(warningResults)
              .replace(regexMatcher, '')
              .replace(fixableMessageMatcher, '')
          );
        }

        for (const { output } of results) {
          if (output && fix) {
            fixes.push(ctx.file.text(file.path, output));
          }
        }
      });

      return fixes;
    });
  }
);

function filterResultsBySeverity(
  results: CLIEngine.LintResult[],
  severity: SeverityLevel
) {
  const filtered: CLIEngine.LintResult[] = [];

  const count = (sevLevel: SeverityLevel, messagesSum: number) => {
    return severity === sevLevel ? messagesSum : 0;
  };

  results.forEach(result => {
    const filteredMessages = result.messages.filter(
      mess => mess.severity === severity
    );

    if (filteredMessages.length > 0) {
      filtered.push({
        ...result,
        messages: filteredMessages,
        errorCount: count(SeverityLevel.Error, filteredMessages.length),
        warningCount: count(SeverityLevel.Warning, filteredMessages.length),
        fixableErrorCount: result.fixableErrorCount,
        fixableWarningCount: 0
      });
    }
  });

  return filtered;
}
