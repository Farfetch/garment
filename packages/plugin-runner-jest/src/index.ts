import {
  Context,
  defineRunner,
  defineWatch,
  Batch,
  defineOptionsFromJSONSchema
} from '@garment/runner';
import * as fs from 'fs-extra';
import * as Path from 'path';
import * as tempy from 'tempy';
import { Config } from '@jest/types';

const jestCLI = require('jest');

/**
 * Runs Jest
 * @example {
    "runner": "jest",
    "options": {
      "jestConfig": "jest.config.js"
    }
  }
 */
export interface BaseJestRunnerOpts {
  /**
   * The config file to load Jest's config from.
   * @format path
   * @default jest.config.js
   */
  configFile: string;

  testPathPattern?: string;
  updateSnapshot?: boolean;
  u?: boolean;
  _?: string | boolean;
}

export interface JestRunnerOptions extends BaseJestRunnerOpts {
  [key: string]: string | number | boolean | undefined;
}

export default defineRunner(
  defineOptionsFromJSONSchema<JestRunnerOptions>(require('./schema.json')),
  async ctx => {
    process.env.NODE_ENV = 'test';
    return jestCLI.run(await getJestArgs(ctx));
  }
);

export const watch = defineWatch(
  defineOptionsFromJSONSchema<JestRunnerOptions>(require('./schema.json')),
  async ctx => {
    process.env.NODE_ENV = 'test';
    return jestCLI.run([...(await getJestArgs(ctx)), '--watch']);
  }
);

async function getJestArgs(ctx: Context<JestRunnerOptions>) {
  const opts = ctx.options;
  const {
    _,
    testPathPattern = _,
    u,
    updateSnapshot = u,
    outputDir,
    configFile,
    ...rest
  } = opts;

  const jestConfig = await getConfig(ctx);

  const outputFilePath = tempy.file({ extension: 'json' });
  await fs.outputJSON(outputFilePath, jestConfig);

  const args = ['--config', outputFilePath];
  if (testPathPattern && testPathPattern !== true) {
    args.push('--testPathPattern', String(testPathPattern));
  }
  if (updateSnapshot) {
    args.push('--updateSnapshot');
  }

  const restArgs = Object.entries(rest)
    .map(([key, value]) => [`--${key}`, String(value)])
    .reduce((a, b) => [...a, ...b], []);

  args.push(...restArgs);

  ctx.logger.silly('jest config:\n', JSON.stringify(jestConfig, null, 2));
  ctx.logger.silly('jest args:\n', JSON.stringify(args, null, 2));
  ctx.logger.silly('temp config path:\n', outputFilePath);

  return args;
}

export async function getConfig(ctx: Context<JestRunnerOptions>) {
  const batch = [...ctx.batch()];

  const projsByConfig: {
    [key: string]: Batch<JestRunnerOptions>[];
  } = {};

  batch.forEach(item => {
    const { configFile } = item.options;

    if (!projsByConfig[configFile]) {
      projsByConfig[configFile] = [];
    }
    projsByConfig[configFile].push(item);
  });

  const projsKeys = Object.keys(projsByConfig);
  const isMultipleConfigs = projsKeys.length > 1;
  const jestConfig: Partial<Config.DefaultOptions> = {
    rootDir: ctx.workspace.cwd,
    projects: []
  };

  const configEntries = Object.entries(projsByConfig).map(
    ([configPath, items]) =>
      [
        configPath,
        require(configPath) as Partial<
          Config.DefaultOptions & Config.ProjectConfig
        >,
        items
      ] as const
  );

  const hasCollectCoverage = configEntries.some(([, config]) =>
    Boolean(config.collectCoverage)
  );

  for (const [configPath, projectConfig, items] of configEntries) {
    const rootDir = projectConfig.rootDir || Path.dirname(configPath);
    const roots = items.map(
      item => '<rootDir>/' + Path.relative(rootDir, item.project.fullPath)
    );
    projectConfig.rootDir = rootDir;
    projectConfig.roots = roots;

    if (isMultipleConfigs) {
      if (projectConfig.collectCoverage) {
        delete projectConfig.collectCoverage;
      } else if (hasCollectCoverage) {
        projectConfig.coveragePathIgnorePatterns = roots;
      }
    }

    if (isMultipleConfigs && jestConfig.projects) {
      if (projectConfig.projects && projectConfig.projects.length > 0) {
        const newProjects = projectConfig.projects as Config.ProjectConfig[];

        newProjects.forEach(project => {
          if (projectConfig.rootDir) {
            project.rootDir = projectConfig.rootDir;
          }
        });

        jestConfig.projects.push(...newProjects);
      } else {
        jestConfig.projects.push(projectConfig as Config.ProjectConfig);
      }
    } else {
      Object.assign(jestConfig, projectConfig);
    }
  }

  if (hasCollectCoverage) {
    jestConfig.collectCoverage = true;
  }

  return jestConfig;
}
