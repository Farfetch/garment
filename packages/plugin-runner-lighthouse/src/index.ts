import { defineRunner, defineOptionsFromJSONSchema } from '@garment/runner';
const lighthouse = require('lighthouse');
import * as chromeLauncher from 'chrome-launcher';
import dargs = require('dargs');

/**
 * Runs Lighthouse
 * @example {
    "runner": "lighthouse",
    "output": "{{projectDir}}/lighthouse-output",
    "options": {
      "port": 3000,
      "budget": "{{projectDir}}/budget.json",
      "scoreThreshold": {
        "performance": 90,
        "accessibility": 90,
        "seo": 100,
        "pwa": 80,
        "best-practices": 80
      }
    }
  }
 */
interface LighthouseRunnerOptions {
  url?: string;
  port?: number;
  chromeFlags?: any;

  /**
   * @format path
   */
  budget: string;
  scoreThreshold?: {
    [key: string]: number;
  };

  /**
   * @format path
   */
  config?: string;
  reportFormat?: 'html' | 'json';
}

export default defineRunner(
  defineOptionsFromJSONSchema<LighthouseRunnerOptions>(
    require('./schema.json')
  ),
  async ctx => {
    const { logger, options, fs, project } = ctx;
    const {
      port,
      url = `http://localhost:${port}`,
      budget,
      scoreThreshold,
      config,
      chromeFlags,
      reportFormat = 'html'
    } = options;

    async function launchChromeAndRunLighthouse(url: string, flags: any) {
      const chrome = await chromeLauncher.launch({
        chromeFlags: flags.chromeFlags
      });
      flags.port = chrome.port;
      const results = await lighthouse(url, flags, flags.config);
      await chrome.kill();
      return results;
    }

    function checkBudget(lhr: any) {
      const categories = Object.entries(lhr.categories);

      if (lhr.audits['performance-budget']) {
        logger.info(`${lhr.audits['performance-budget'].title}\n`);

        lhr.audits['performance-budget'].details.items.forEach((item: any) => {
          logger.info(item.label);
          item.sizeOverBudget
            ? logger.error(
                `Exceeds budget by:${formatBytes(item.sizeOverBudget)}\n`
              )
            : logger.success(`${formatBytes(item.size)}\n`);
        });

        if (scoreThreshold) {
          logger.info('Score Thresholds:\n');
          categories.map(cat => {
            const categorydata: any = cat[1];
            const score = categorydata.score * 100;
            const threshold = scoreThreshold[categorydata.id];
            if (threshold) {
              const message = `Score: ${score} Threshold: ${threshold}\n`;
              logger.info(`${categorydata.title}`);
              score >= threshold
                ? logger.success(message)
                : logger.error(message);
            }
          });
        }
      }
      logAuditResults(categories, lhr);
    }

    function logAuditResults(categories: any[], lhr: any) {
      logger.info(`Audit Results:\n`);
      categories.map(item => {
        const categorydata: any = item[1];
        logger.info(`${categorydata.title}\n`);

        categorydata.auditRefs.forEach((ref: any) => {
          if (lhr.audits[ref.id].score === 1) {
            logger.success(`${lhr.audits[ref.id].title}\n`);
          }
        });

        categorydata.auditRefs.forEach((ref: any) => {
          if (lhr.audits[ref.id].score > 0 && lhr.audits[ref.id].score < 1) {
            logger.warn(`${lhr.audits[ref.id].title}\n`);
          }
        });

        categorydata.auditRefs.forEach((ref: any) => {
          if (lhr.audits[ref.id].score === 0) {
            logger.error(`${lhr.audits[ref.id].title}\n`);
          }
        });
      });

      logger.info(
        `You can get more info by using the generated lighthouse.json file at https://googlechrome.github.io/lighthouse/viewer/ \n`
      );
    }

    function formatBytes(bytes: number, decimals = 2) {
      if (bytes === 0) return '0 Bytes';
      const k = 1024;
      const dm = decimals < 0 ? 0 : decimals;
      const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];
      const i = Math.floor(Math.log(bytes) / Math.log(k));

      return parseFloat((bytes / Math.pow(k, i)).toFixed(dm)) + ' ' + sizes[i];
    }

    let flags: any = {};

    if (budget) {
      flags.budgets = JSON.parse(fs.readFileSync(budget, { encoding: 'utf8' }));
    }

    if (config) {
      flags.config = JSON.parse(fs.readFileSync(config, { encoding: 'utf8' }));
    }

    if (chromeFlags) {
      flags.chromeFlags = dargs(chromeFlags);
    }

    flags.output = reportFormat;

    const results = await launchChromeAndRunLighthouse(url, flags);
    checkBudget(results.lhr);
    return ctx.file.text(`${project.fullPath}/lighthouse.html`, results.report);
  }
);
