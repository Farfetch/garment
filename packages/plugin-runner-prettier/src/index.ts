import {
  defineOptionsFromJSONSchema,
  defineRunner,
  File,
} from '@garment/runner';
import * as prettier from 'prettier';

interface PrettierRunnerOptions {
  /**
   * The config file to load Prettier's configuration from.
   * @format path
   */
  configFile?: string;

  /**
   * Fix automatically linting errors.
   */
  fix?: boolean;
}

const runnerOptions = defineOptionsFromJSONSchema<PrettierRunnerOptions>(
  require('./schema.json')
);

export default defineRunner(runnerOptions, async (ctx) => {
  const { logger } = ctx;
  const { configFile = '', outputDir, fix: fixFlag } = ctx.options;
  const fix = Boolean(outputDir) && fixFlag;

  logger.info(`Linting files with Prettier`);
  ctx.input(async (files) => {
    const fixes: File[] = [];

    const prettierConfig = await prettier.resolveConfig(configFile);
    if (!prettierConfig) {
      logger.info(`Prettier config not found`);
      return;
    }

    files.forEach((file) => {
      //TODO What to do about errors like this?
      // error SyntaxError: Unexpected token, expected ";" (2:11)
      const isChecked = prettier.check(file.data.toString('utf8'), {
        ...prettierConfig,
        filepath: file.absolutePath,
      });

      if (!isChecked && fix) {
        logger.info(`Going to format`);
        logger.warn(`output`, file.data.toString('utf8'));
        const output = prettier.format(file.data.toString('utf8'), {
          ...prettierConfig,
          filepath: file.absolutePath,
        });

        fixes.push(ctx.file.text(file.path, output));
      }
    });

    return fixes;
  });
});
