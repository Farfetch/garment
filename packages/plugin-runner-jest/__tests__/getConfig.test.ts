import {
  createBatchContext,
  initFixtureHelper,
  replaceTestPath
} from '@garment/fixture-helper';
import { getConfig, JestRunnerOptions } from '../src';

const { initFixture, clean } = initFixtureHelper(module, {
  tempDir: __dirname + 'tmp__'
});

afterAll(clean);

test('Handles multiple projects sharing the same config', async () => {
  const cwd = await initFixture('basic');
  const { context } = await createBatchContext<JestRunnerOptions>(
    {
      foo: { configFile: 'jest.config.js' },
      bar: { configFile: 'jest.config.js' }
    },
    { cwd, schema: require('../src/schema.json') }
  );

  const jestConfig = await getConfig(context);

  expect(replaceTestPath(jestConfig, cwd)).toMatchSnapshot();
});

test('Handles multiple projects sharing the same config and one having different config', async () => {
  const cwd = await initFixture('basic');
  const { context } = await createBatchContext<JestRunnerOptions>(
    {
      foo: { configFile: 'jest.config.js' },
      bar: { configFile: 'jest.config.js' },
      test: { configFile: '{{projectDir}}/jest.config.js' }
    },
    { cwd, schema: require('../src/schema.json') }
  );

  const jestConfig = await getConfig(context);

  expect(replaceTestPath(jestConfig, cwd)).toMatchSnapshot();
});
