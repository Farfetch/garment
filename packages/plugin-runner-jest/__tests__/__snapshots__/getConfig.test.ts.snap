// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handles multiple projects sharing the same config 1`] = `
Object {
  "collectCoverage": true,
  "projects": Array [],
  "rootDir": "/test_path",
  "roots": Array [
    "<rootDir>/foo",
    "<rootDir>/bar",
  ],
  "testEnvironment": "jsdom",
}
`;

exports[`Handles multiple projects sharing the same config and one having different config 1`] = `
Object {
  "collectCoverage": true,
  "projects": Array [
    Object {
      "rootDir": "/test_path",
      "roots": Array [
        "<rootDir>/foo",
        "<rootDir>/bar",
      ],
      "testEnvironment": "jsdom",
    },
    Object {
      "coveragePathIgnorePatterns": Array [
        "<rootDir>/",
      ],
      "rootDir": "/test_path/test",
      "roots": Array [
        "<rootDir>/",
      ],
      "testEnvironment": "node",
    },
  ],
  "rootDir": "/test_path",
}
`;
